File: packages/metadata_core/lib/src/video_meta.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
import 'field.dart';

bool rotationSwapsAxes(int? rotation) =>
    rotation == 90 || rotation == 270;

class VideoMeta {
  const VideoMeta({
    required this.durationMs,
    required this.widthRaw,
    required this.heightRaw,
    required this.rotation,
    required this.creationTime,
    required this.locationLatitude,
    required this.locationLongitude,
    required this.codec,
    required this.fileSizeBytes,
    required this.mime,
  });

  final MetaField<int> durationMs;
  final MetaField<int> widthRaw;
  final MetaField<int> heightRaw;
  final MetaField<int> rotation;
  final MetaField<String> creationTime;
  final MetaField<double> locationLatitude;
  final MetaField<double> locationLongitude;
  final MetaField<String> codec;
  final MetaField<int> fileSizeBytes;
  final MetaField<String> mime;

  int? get width {
    final int? rawWidth = widthRaw.value;
    final int? rawHeight = heightRaw.value;
    if (rawWidth == null || rawHeight == null) {
      return null;
    }
    return rotationSwapsAxes(rotation.value) ? rawHeight : rawWidth;
  }

  int? get height {
    final int? rawWidth = widthRaw.value;
    final int? rawHeight = heightRaw.value;
    if (rawWidth == null || rawHeight == null) {
      return null;
    }
    return rotationSwapsAxes(rotation.value) ? rawWidth : rawHeight;
  }
}

File: native/vision-core/src/main/kotlin/com/visionengine/core/VideoMeta.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
package com.visionengine.core

fun rotationSwapsAxes(rotation: Long?): Boolean =
    rotation == 90L || rotation == 270L

data class VideoMeta(
    val durationMs: MetaField<Long>,
    val widthRaw: MetaField<Long>,
    val heightRaw: MetaField<Long>,
    val rotation: MetaField<Long>,
    val creationTime: MetaField<String>,
    val locationLatitude: MetaField<Double>,
    val locationLongitude: MetaField<Double>,
    val codec: MetaField<String>,
    val fileSizeBytes: MetaField<Long>,
    val mime: MetaField<String>,
) {
    val width: Long?
        get() {
            val rawWidth = widthRaw.value
            val rawHeight = heightRaw.value
            if (rawWidth == null || rawHeight == null) {
                return null
            }
            return if (rotationSwapsAxes(rotation.value)) rawHeight else rawWidth
        }

    val height: Long?
        get() {
            val rawWidth = widthRaw.value
            val rawHeight = heightRaw.value
            if (rawWidth == null || rawHeight == null) {
                return null
            }
            return if (rotationSwapsAxes(rotation.value)) rawWidth else rawHeight
        }
}
